crates/matrix-sdk/src/sliding_sync/mod.rs

```diff
@@ -819,14 +824,54 @@ impl StickyData for SlidingSyncStickyParameters {
     }
 }
 
+/// As of 2023-07-13, the sliding sync proxy doesn't provide us with `limited`
+/// correctly, so we cheat and "correct" it using heuristics here.
+/// TODO remove this workaround as soon as support of the `limited` flag is
+/// properly implemented in the open-source proxy: https://github.com/matrix-org/sliding-sync/issues/197
+fn compute_limited(
+    known_rooms: &BTreeMap<OwnedRoomId, SlidingSyncRoom>,
+    response_rooms: &mut BTreeMap<OwnedRoomId, v4::SlidingSyncRoom>,
+) {
+    for (room_id, room) in response_rooms {
+        // Only rooms marked as initially loaded are subject to the fixup.
+        let initial = room.initial.unwrap_or(false);
+        if !initial {
+            continue;
+        }
+
+        // If the known room had some timeline events, consider it's a `limited` if
+        // there's absolutely no overlap between the known events and
+        // the new events in the timeline.
+        if let Some(known_room) = known_rooms.get(room_id) {
+            // Gather all the known event IDs. Ignore events that don't have an event ID.
+            let known_events: HashSet<OwnedEventId> = HashSet::from_iter(
+                known_room.timeline_queue().into_iter().filter_map(|event| event.event_id()),
+            );
+
+            // There's overlap if, and only if, there's at least one event in the
+            // response's timeline that matches an event id we've seen before.
+            let overlap = room.timeline.iter().any(|seen_event| {
+                seen_event
+                    .deserialize()
```

Comment:
Fully deserializing the event is a rather inefficient way of getting the event ID. See `matrix_sdk_common::deserialized_responses::SyncTimelineEvent::event_id`.

Reply:
Ah nice, thanks, it's indeed better.